File: Exercise_1.cpp
/* 
    Time Complexity                              :  O(2*N) ~ O(N). In the worst case, push all the n-1 elements in the stack and then n th temperature is the highest. 
                                                    So pop everything out from the stack operation is going to be done.
    Space Complexity                             :  O(N) - Store the n-1 indices in the stack.
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/daily-temperatures/submissions/


class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> st;
        st.push(0);
        int n = temperatures.size();
        vector<int> ans(n,0);
       for(int i=1;i<n;i++) {
           int cur_temp = temperatures[i];
           while(!st.empty() and cur_temp > temperatures[st.top()]) {
                ans[st.top()] = i - st.top();
                st.pop();
           }
           st.push(i);
       }
        return ans;
    }
};

File: Exercise_1.py
# Problem1 Daily Temperatures (https://leetcode.com/problems/daily-temperatures/)
# // Time Complexity :O(2N) = O(N)
# // Space Complexity :O(N)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No
class Solution:
    def dailyTemperatures(self, temperatures):
        if not temperatures:
            return []
        result = [0] * len(temperatures)
        stack = []
        for i in range(len(temperatures)):
            while stack and temperatures[stack[-1]] < temperatures[i]:
                popped_index = stack.pop()
                result[popped_index] = i - popped_index
            stack.append(i)
        return result


sol = Solution()
print(sol.dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]))
